File: proj_2/app/src/main/java/com/example/proj_2/PermissionActivity.java
package com.example.proj_2;

import android.Manifest;
import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.content.pm.PackageManager;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.Toast;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AlertDialog;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import com.google.android.material.snackbar.Snackbar;

import java.util.ArrayList;
import java.util.List;


public class PermissionActivity extends AppCompatActivity {
    private static final int PERMISSIONS_REQUEST_CODE = 100;

    private int MY_PERMISSION_REQUEST_SMS = 1001;
    String[] REQUIRED_PERMISSIONS = { Manifest.permission.READ_EXTERNAL_STORAGE,
            Manifest.permission.READ_CONTACTS, Manifest.permission.CALL_PHONE,
            Manifest.permission.CAMERA,  Manifest.permission.READ_PHONE_STATE,
            Manifest.permission.WRITE_CONTACTS, Manifest.permission.SEND_SMS,
            Manifest.permission.READ_SMS, Manifest.permission.WRITE_EXTERNAL_STORAGE,
            Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION };
    public int[] grandResults = {-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1 };

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        // Require Permission
        GetPermission();
        onRequestPermissionsResult(PERMISSIONS_REQUEST_CODE, REQUIRED_PERMISSIONS, grandResults);

        super.onCreate(savedInstanceState);
        startMainActivity();

    }
    private void startMainActivity(){
        Intent intent = new Intent(this, facebookloginActivity.class);
        startActivity(intent);
        finish();
    }

    public void GetPermission() {
        LinearLayout mLayout = (LinearLayout)findViewById(R.id.activity_permission);
        int readExternalStoragePermission = ContextCompat.checkSelfPermission(this, REQUIRED_PERMISSIONS[0]);
        int readContactPermission = ContextCompat.checkSelfPermission(this, REQUIRED_PERMISSIONS[1]);
        int readCallPermission = ContextCompat.checkSelfPermission(this, REQUIRED_PERMISSIONS[2]);
        int cameraPermission = ContextCompat.checkSelfPermission(this, REQUIRED_PERMISSIONS[3]);
        int statePermission = ContextCompat.checkSelfPermission(this, REQUIRED_PERMISSIONS[4]);
        int writeContactPermission = ContextCompat.checkSelfPermission(this, REQUIRED_PERMISSIONS[5]);
        int sendSMSPermission =  ContextCompat.checkSelfPermission(this, REQUIRED_PERMISSIONS[6]);
        int readSMSPermission = ContextCompat.checkSelfPermission(this, REQUIRED_PERMISSIONS[7]);
        int writeExternalPermission = ContextCompat.checkSelfPermission(this, REQUIRED_PERMISSIONS[8]);

        int accessFineLoaction = ContextCompat.checkSelfPermission(this, REQUIRED_PERMISSIONS[9]);
        int accessCoarseLocation = ContextCompat.checkSelfPermission(this, REQUIRED_PERMISSIONS[10]);

        ActivityCompat.requestPermissions(PermissionActivity.this, new String[] {Manifest.permission.SEND_SMS}, MY_PERMISSION_REQUEST_SMS);


        grandResults[0] = readExternalStoragePermission;
        grandResults[1] = readContactPermission;
        grandResults[2] = readCallPermission;
        grandResults[3] = cameraPermission;
        grandResults[4] = statePermission;
        grandResults[5] = writeContactPermission;
        grandResults[6] = sendSMSPermission;
        grandResults[7] = readSMSPermission;
        grandResults[8] = writeExternalPermission;
        grandResults[9] = accessFineLoaction;
        grandResults[10] = accessCoarseLocation;


        if (!(grandResults[0] == PackageManager.PERMISSION_GRANTED
                && grandResults[1] == PackageManager.PERMISSION_GRANTED
                && grandResults[2] == PackageManager.PERMISSION_GRANTED
                && grandResults[3] == PackageManager.PERMISSION_GRANTED
                && grandResults[4] == PackageManager.PERMISSION_GRANTED
                && grandResults[5] == PackageManager.PERMISSION_GRANTED
                && grandResults[6] == PackageManager.PERMISSION_GRANTED
                && grandResults[7] == PackageManager.PERMISSION_GRANTED
                && grandResults[8] == PackageManager.PERMISSION_GRANTED
                && grandResults[9] == PackageManager.PERMISSION_GRANTED
                && grandResults[10] == PackageManager.PERMISSION_GRANTED)) {


            if (ActivityCompat.shouldShowRequestPermissionRationale(this, REQUIRED_PERMISSIONS[0])
                    || ActivityCompat.shouldShowRequestPermissionRationale(this, REQUIRED_PERMISSIONS[1])
                    || ActivityCompat.shouldShowRequestPermissionRationale(this, REQUIRED_PERMISSIONS[2])
                    || ActivityCompat.shouldShowRequestPermissionRationale(this, REQUIRED_PERMISSIONS[3])
                    || ActivityCompat.shouldShowRequestPermissionRationale(this, REQUIRED_PERMISSIONS[4])
                    || ActivityCompat.shouldShowRequestPermissionRationale(this, REQUIRED_PERMISSIONS[5])
                    || ActivityCompat.shouldShowRequestPermissionRationale(this, REQUIRED_PERMISSIONS[6])
                    || ActivityCompat.shouldShowRequestPermissionRationale(this, REQUIRED_PERMISSIONS[7])
                    || ActivityCompat.shouldShowRequestPermissionRationale(this, REQUIRED_PERMISSIONS[8])
                    || ActivityCompat.shouldShowRequestPermissionRationale(this, REQUIRED_PERMISSIONS[9])
                    || ActivityCompat.shouldShowRequestPermissionRationale(this, REQUIRED_PERMISSIONS[10])) {
                Snackbar.make(mLayout, "이 앱을 실행하려면 외부 저장소, 연락처, 전화 접근 권한이 필요합니다.",
                        Snackbar.LENGTH_INDEFINITE).setAction("확인", new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        // 3-3. 사용자에게 퍼미션 요청을 합니다. 요청 결과는 onRequestPermissionResult에서 수신됩니다.
                        ActivityCompat.requestPermissions(PermissionActivity.this, REQUIRED_PERMISSIONS,
                                PERMISSIONS_REQUEST_CODE);
                    }
                }).show();
            } else {
                // 4-1. 사용자가 퍼미션 거부를 한 적이 없는 경우에는 퍼미션 요청을 바로 합니다.
                // 요청 결과는 onRequestPermissionResult에서 수신됩니다.
                ActivityCompat.requestPermissions(this, REQUIRED_PERMISSIONS,
                        PERMISSIONS_REQUEST_CODE);
            }
        }
    }
}

File: sj/project1_final/app/src/main/java/com/example/project1_final/PermissionActivity.kt
package com.example.project1_final

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.util.Log
import android.widget.Toast
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat

class PermissionActivity: AppCompatActivity() {
    companion object {
        const val TAG = "PERMISSION_ACTIVITY"
        const val PERMISSION_REQUEST_CODE = 1001
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        //setContentView(R.layout.activity_permission)

        val permissionList = arrayOf(
                Manifest.permission.READ_EXTERNAL_STORAGE,
                Manifest.permission.WRITE_EXTERNAL_STORAGE,
                Manifest.permission.READ_CONTACTS,
                Manifest.permission.WRITE_CONTACTS
        )

        if (checkPermission(permissionList).isEmpty()
        ) {
            Toast.makeText(this, "permission granted", Toast.LENGTH_SHORT).show()
            startMainActivity()
        } else {
            requestRuntimePermissions(permissionList)
        }


    }

    private fun checkPermission(permissions: Array<String>): MutableList<String> {
        val notGrantedArray = mutableListOf<String>()

        permissions.forEach {
            if (ContextCompat.checkSelfPermission(
                    this,
                    it
                ) != PackageManager.PERMISSION_GRANTED){
                notGrantedArray.add(it)
            }
        }

        return notGrantedArray
    }

    private fun startMainActivity(){
        val intent: Intent = Intent(this, MainActivity::class.java)
        startActivity(intent)
        finish()
    }

    private fun requestRuntimePermissions(permissions: Array<String>) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            // Don't need to request
            // It's supported in devices running VERSION_CODES.M or higher
            return
        }

        requestPermissions(
            permissions,
            PermissionActivity.PERMISSION_REQUEST_CODE
        )
    }

    @RequiresApi(Build.VERSION_CODES.M)
    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>,
                                            grantResults: IntArray) {
        when (requestCode) {
            PermissionActivity.PERMISSION_REQUEST_CODE -> {
                if (grantResults.isEmpty()) {
                    throw RuntimeException("Empty permission result")
                }
                if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    //showDialog("Permission granted")
                    startMainActivity()
                } else {
                    if (shouldShowRequestPermissionRationale(
                            Manifest.permission.READ_EXTERNAL_STORAGE)) {
                        Log.d(PermissionActivity.TAG, "User declined, but i can still ask for more")
                        requestPermissions(
                            permissions,
                            PermissionActivity.PERMISSION_REQUEST_CODE
                        )
                    } else {
                        Log.d(PermissionActivity.TAG, "User declined and i can't ask")
                        showDialogToGetPermission()
                    }
                }
            }
        }
    }

    private fun showDialog(msg: String) {
        val builder = AlertDialog.Builder(this)
        builder.setTitle("Dialog")
            .setMessage(msg)
        val dialog = builder.create()
        dialog.show()
    }

    private fun showDialogToGetPermission() {
        val builder = AlertDialog.Builder(this)
        builder.setTitle("Permisisons request")
            .setMessage("We need the location permission for some reason. " +
                    "You need to move on Settings to grant some permissions")
        builder.setPositiveButton("OK") { _, _ ->
            val intent = Intent(
                Settings.ACTION_APPLICATION_DETAILS_SETTINGS,
                Uri.fromParts("package", packageName, null))
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            startActivity(intent)
        }
        builder.setNegativeButton("Later") { _, _ ->
            // ignore
        }
        val dialog = builder.create()
        dialog.show()
    }
}